qiskit/visualization/timeline/drawing_objects.py

```diff
@@ -0,0 +1,285 @@
+It should be noted that there will be no unittest for a plotter interface, which takes
+drawing objects and output an image data, we should avoid adding a complicated data structure
+that has a context of the scheduled circuit program.
+
+Usually a drawing object is associated to the specific bit. `BitLinkData` is the
+exception of this framework because it takes multiple bits to connect.
+Position of the link is dynamically updated according to the user preference of bit to show.
+While other objects are general format to represent specific shapes, the `BitLinkData` is
+only used by the bit links and thus has no `data_type` input.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Optional, Dict, Any, List
+
+from qiskit.visualization.timeline import types
+
+
+class ElementaryData(ABC):
+    """Base class of the scheduled circuit visualization object.
+
+    Note that drawing objects are mutable.
+    """
+    __hash__ = None
+
+    def __init__(self,
+                 data_type: str,
+                 meta: Optional[Dict[str, Any]],
+                 visible: bool,
+                 styles: Optional[Dict[str, Any]]):
+        """Create new drawing object.
+
+        Args:
+            data_type: String representation of this drawing object.
+            meta: Meta data dictionary of the object.
+            visible: Set ``True`` to show the component on the canvas.
+            styles: Style keyword args of the object. This conforms to `matplotlib`.
+        """
+        self.data_type = data_type
+        self.meta = meta
+        self.visible = visible
+        self.styles = styles
+
+    @property
+    @abstractmethod
+    def data_key(self):
+        """Return unique hash of this object."""
+        pass
+
+    def __repr__(self):
+        return "{}(type={}, key={})".format(self.__class__.__name__,
+                                            self.data_type,
+                                            self.data_key)
+
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and self.data_key == other.data_key
+
+
+class LineData(ElementaryData):
+    """Drawing object that represents line shape."""
+    def __init__(self,
+                 data_type: str,
+                 bit: types.Bits,
+                 x: List[types.Coordinate],
+                 y: List[types.Coordinate],
+                 meta: Dict[str, Any] = None,
+                 visible: bool = True,
+                 styles: Dict[str, Any] = None):
+        """Create new line.
+
+        Args:
+            data_type: String representation of this drawing object.
+            bit: Bit associated to this object.
+            x: Horizontal coordinate sequence of this line.
+            y: Vertical coordinate sequence of this line.
+            meta: Meta data dictionary of the object.
+            visible: Set ``True`` to show the component on the canvas.
+            styles: Style keyword args of the object. This conforms to `matplotlib`.
```

Comment:
Other backends should work, too, right? Should this not be
```suggestion
            styles: Style keyword args of the object. This conforms to the plotter interface.
```
?

Reply:
No, this is literary conforms to the matplotlib format, i.e. kwarg name is based on matplotlib.

Reply:
So, you would need a new `LineData` object to support a different plotter? Or is there somewhere that styles are converted to matplotlib kwargs to the plotter kwargs? I didn't see that anywhere

Reply:
Exactly. As the initial step I'm going to introduce matplotlib-based plotter, but in future we may introduce some other plotters like bokeh. Then we need to write some kwarg converter for that plotter. At this moment I don't plan to make some general stylesheet framework.